src/any_agent/frameworks/tinyagent.py

```diff
@@ -209,199 +181,47 @@ async def _run_async(self, prompt: str, **kwargs: Any) -> str:
-                tool_args = {}
-
-                if tool_call.function.arguments:
-                    tool_args = json.loads(tool_call.function.arguments)
-                    logger.debug("Tool arguments: %s", tool_args)
-
-                tool_message = {
-                    "role": "tool",
-                    "tool_call_id": tool_call.id,
-                    "content": "",
-                    "name": tool_name,
-                }
-
-                # Check if tool is in exit loop tools
-                exit_tools = options["exit_loop_tools"]
-                if exit_tools and tool_name in [
-                    t["function"]["name"] for t in exit_tools
-                ]:
-                    logger.debug("Exit tool called: %s", tool_name)
-                    exit_tool_called = True
-                    self.messages.append(tool_message)
-                    combined_results.append(str(tool_message["content"]))
-                    continue
-
-                # Check if the tool exists
-                if tool_name not in self.clients:
-                    logger.error("Tool %s not found in registered tools", tool_name)
-                    tool_message["content"] = (
-                        f"Error: No tool found with name: {tool_name}"
-                    )
-                else:
-                    client = self.clients[tool_name]
-                    try:
-                        result = await client.call_tool(
-                            {"name": tool_name, "arguments": tool_args}
-                        )
-                        if (
-                            isinstance(result, dict)
-                            and "content" in result
-                            and isinstance(result["content"], list)
-                        ):
-                            tool_message["content"] = result["content"][0]["text"]
-                        else:
-                            tool_message["content"] = str(result)
-
-                        logger.debug(
-                            "Tool result: %s...",
-                            tool_message["content"][:50]
-                            if tool_message["content"]
-                            else "Empty",
-                        )
-                    except Exception as e:
-                        logger.error("Error calling tool %s: %s", tool_name, e)
-                        tool_message["content"] = f"Error calling tool {tool_name}: {e}"
-
-                self.messages.append(tool_message)
-                combined_results.append(str(tool_message["content"]))
-
-            # If an exit tool was called, return early with the combined results
-            if exit_tool_called:
-                return "\n".join(combined_results)
+                return str(current_last["content"])
 
-            return "\n".join(combined_results)
-        return str(message.content)
+        return "Max turns reached"
```

Comment:
Hmm. Can this throw an exception instead of returning? I think that's what the other frameworks would do. Might be weird to have it successfully complete but have this error. 

We have the ticket https://github.com/mozilla-ai/any-agent/issues/368 for a refactor of how we handle failures.